src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java

```diff
@@ -40,31 +52,203 @@ public Optional<SerializingMutator<?>> tryCreate(AnnotatedType type, MutatorFact
+          type, Float.NEGATIVE_INFINITY, Float.POSITIVE_INFINITY, true, mathFunctions));
     } else if (clazz == double.class || clazz == Double.class) {
-      return Optional.of(new DoubleMutator(type, -Double.MAX_VALUE, Double.MAX_VALUE));
+      return Optional.of(new DoubleMutator(
+          type, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, true, mathFunctions));
     } else {
       return Optional.empty();
     }
   }
 
-  private static final class FloatMutator extends SerializingMutator<Float> {
-    FloatMutator(AnnotatedType type, Float defaultMinValueForType, Float defaultMaxValueForType) {}
+  static final class FloatMutator extends SerializingMutator<Float> {
+    private static final int MANTISSA_RANDOM_WALK_RANGE = 1000;
+    private static final int EXPONENT_RANDOM_WALK_RANGE = Float.MAX_EXPONENT;
+    private static final int INV_CHANCE_SPECIAL_VALUE = 1000;
+    private final float minValue;
+    private final float maxValue;
+    private final boolean allowNaN;
+    private final float[] specialValues;
+    private ArrayList<Function<Double, Double>> mathFunctions;
+
+    FloatMutator(AnnotatedType type, Float defaultMinValueForType, Float defaultMaxValueForType,
+        boolean defaultAllowNaN, ArrayList<Function<Double, Double>> mathFunctions) {
+      this.mathFunctions = mathFunctions;
+      float minValue = defaultMinValueForType;
+      float maxValue = defaultMaxValueForType;
+      boolean allowNaN = defaultAllowNaN;
+      // InRange is not repeatable, so the loop body will apply exactly once.
+      for (Annotation annotation : type.getAnnotations()) {
+        if (annotation instanceof FloatInRange) {
+          FloatInRange floatInRange = (FloatInRange) annotation;
+          minValue = floatInRange.min();
+          maxValue = floatInRange.max();
+          allowNaN = floatInRange.allowNaN();
+        }
+      }
+
+      require(minValue <= maxValue,
+          format("[%f, %f] is not a valid interval: %s", minValue, maxValue, type));
+      require(minValue != maxValue,
+          format(
+              "[%f, %f] can not be mutated, use a constant instead: %s", minValue, maxValue, type));
+      this.minValue = minValue;
+      this.maxValue = maxValue;
+      this.allowNaN = allowNaN;
+      this.specialValues = collectSpecialValues(minValue, maxValue);
+    }
+
+    private float[] collectSpecialValues(float minValue, float maxValue) {
+      float[] specialValues = new float[] {Float.NEGATIVE_INFINITY, Float.POSITIVE_INFINITY, 0.0f,
+          -0.0f, Float.NaN, Float.MAX_VALUE, Float.MIN_VALUE, -Float.MAX_VALUE, -Float.MIN_VALUE,
+          this.minValue, this.maxValue};
+      ArrayList<Float> specialValuesList = new ArrayList<>();
+      for (float specialValue : specialValues) {
+        if ((allowNaN && Float.isNaN(specialValue))
+            || (specialValue >= minValue && specialValue <= maxValue)) {
+          specialValuesList.add(specialValue);
+        }
+      }
```

Comment:
This part could use `DoubleStream.of` and collect into a `List<Double>` that you then convert to an array using the code below. This would also give you `distinct` and `sorted`.

Reply:
Great, thanks! I was wondering how to do that with floats.